RNA_folding.py

```diff
@@ -0,0 +1,319 @@
+import numpy as np
+import networkx as nx
+import matplotlib.pyplot as plt
+import dimod
+from dwave.system import LeapHybridCQMSampler
+
+"""
+This file contains functions called  to help in the building and processing of models related to RNA folding.
+The models are built and solved in a different file.
+"""
+
+
+def text_to_matrix(file_name, min_loop=2):
+    """ Reads properly formatted RNA text file and returns a matrix of possible hydrogen bonding pairs.
+
+    Args:
+        file_name: String path to text file.
+        min_loop: Integer minimum number of nucleotides between two bonding nucleotides.
+
+    Returns: Numpy matrix of 0's and 1's, where 1 represents a possible bonding pair.
+    """
+
+    # Requires text file of rna data written in same format as examples.
+    with open(file_name) as f:
+        rna = "".join(("".join(line.split()[1:]) for line in f.readlines())).lower()
+
+    # Create a dictionary of all indices where each nucleotide occurs.
+    # Mote: t vs u is mainly preference of database. While u is in rna, t is in dna.
+    index_dict = {'a': [], 'c': [], 'g': [], 't': [], 'u': []}
+
+    # create a dictionary giving list of indices for each nucleotide.
+    for i, nucleotide in enumerate(rna.lower()):
+        index_dict[nucleotide].append(i)
+
+    # List of possible hydrogen bonds for stems.
+    # Recall that 't' is just a stand in for 'u'.
+    hydrogen_bonds = [('a', 't'), ('a', 'u'), ('c', 'g'), ('g', 't'), ('g', 'u')]
+
+    # Create a 0/1 matrix indicated where bonds may occur.
+    bond_matrix = np.zeros((len(rna), len(rna)), dtype=bool)
+    for pair in hydrogen_bonds:
+        for bond in product(index_dict[pair[0]], index_dict[pair[1]]):
+            if abs(bond[0] - bond[1]) > min_loop:
+                bond_matrix[min(bond), max(bond)] = 1
+
+    return bond_matrix
+
+
+def make_stem_dict(bond_matrix, min_stem=3, min_loop=2):
+    """ Takes a matrix of potential hydrogen binding pairs and returns a dictionary of possible stems.
+
+    Each key is a maximal stem (under inclusion) whose value pair is a list of stems weakly contained in the key.
+    Recording stems in this manner allows for faster computations.
+
+    Args:
+        bond_matrix: Numpy matrix of 0's and 1's, where 1 represents a possible bonding pair.
+        min_stem: Integer minimum number of nucleotides between two bonding nucleotides.
+        min_loop: Integer minimum number of nucleotides between two bonding nucleotides.
+
+    Returns: Dictionary of all possible stems with maximal stems as keys.
+    """
+
+    stem_dict = {}
+    n = bond_matrix.shape[0]
+
+    # Iterate through matrix looking for possible stems.
+    for i in range(n - (2 * min_stem + min_loop)):
+        for j in range(i + 2 * min_stem + min_loop - 1, n):
+            if bond_matrix[i, j]:
+                k = 1
+                while bond_matrix[i + k, j - k]:
+                    bond_matrix[i + k, j - k] = False
+                    k += 1
```

Comment:
I guess the content of `bond_matrix` is such that you will never run into a case where you try to access an element with `j - k < 0`?  You might consider reworking that to enforce it logically using a `for` loop so that avoiding the access error is not dependent on the array content:
```suggestion
                for k in range(1, j + 1):
                    if bond_matrix[i + k, j - k]:
                        bond_matrix[i + k, j - k] = False
                    else:
                        break
```

Reply:
So now I'm seeing that `k` is used after the loop also.  That's still valid with this change, but arguably less readable to use `k` outside the `for` loop.  Up to you what to do here, this just jumped out at me as a situation where it's not obvious that you're preventing an array access on a negative index on line 89.

Reply:
I think I would prefer to add a comment about the strictly upper triangular nature of the matrix. For loops are slow and in this case less readable.